app/src/main/java/ch/hikemate/app/model/profile/ProfileRepositoryFirestore.kt

```diff
@@ -0,0 +1,141 @@
+package ch.hikemate.app.model.profile
+
+import android.util.Log
+import com.google.android.gms.tasks.Task
+import com.google.firebase.Firebase
+import com.google.firebase.auth.auth
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.getField
+
+/**
+ * A Firestore implementation of the profile repository.
+ *
+ * @property db The Firestore database.
+ */
+@Suppress("UNCHECKED_CAST")
+class ProfileRepositoryFirestore(private val db: FirebaseFirestore) : ProfileRepository {
+
+  private val collectionPath = "profiles"
+
+  override fun getNewUid(): String {
+    return db.collection(collectionPath).document().id
+  }
+
+  override fun init(onSuccess: () -> Unit) {
+    // Check if the user is logged in
+    // If the user is logged in, call onSuccess
+    Firebase.auth.addAuthStateListener {
+      if (it.currentUser != null) {
+        onSuccess()
+      }
+    }
+  }
+
+  override fun getProfileById(
+      id: String,
+      onSuccess: (Profile) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    // Check that the profile exists
+
+    db.collection(collectionPath).document(id).get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        val profile = task.result?.let { documentToProfile(it) }
+        if (profile != null) {
+          onSuccess(profile)
+        } else {
+          onFailure(Exception("Profile not found"))
+        }
+      } else {
+        task.exception?.let { e ->
+          Log.e("ProfileRepositoryFirestore", "Error getting document", e)
+          onFailure(e)
+        }
+      }
+    }
+  }
+
+  override fun addProfile(profile: Profile, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    getProfileById(
+        profile.id,
+        onSuccess = {
+          // If we found a profile, we can't create a new one with the same id
+          onFailure(Exception("Profile with id ${profile.id} already exists"))
+        },
+        onFailure = {
+          // If we didn't find a profile, we can create a new one
+          // Call updateProfile to create the profile as it is the same implementation
+          updateProfile(profile, onSuccess, onFailure)
+        })
```

Comment:
Fix: Maybe consider creating a function "isProfileExisting" instead of calling onFailure in onSuccess

Reply:
Done!